File: testdata/polyglot/ts/vuln.ts
// Deliberately vulnerable appsec fixture. Every issue is planted. DO NOT fix.
import { execSync } from 'child_process';
import * as fs from 'fs';
import * as crypto from 'crypto';

export function runCommand(userInput: string): void {
  // PLANT(ts-cmdi, min-profile=standard, CWE-78): OS command injection via child_process.execSync with interpolated user input
  execSync(`echo ${userInput}`);
}

export function readSensitiveFile(filename: string): string {
  const baseDir = '/etc';
  // PLANT-GAP: path traversal via concatenated user filename (CWE-22) — caught by no profile; tracked in docs/coverage.md
  const filePath = baseDir + '/' + filename;
  return fs.readFileSync(filePath, 'utf8');
}

export function hashPassword(password: string): string {
  // PLANT(ts-weak-hash, min-profile=max, CWE-327): weak crypto — createHash("md5") on a password (textbook CWE-328; semgrep emits CWE-327)
  return crypto.createHash('md5').update(password).digest('hex');
}

File: testdata/polyglot/python/vuln.py
# Deliberately vulnerable appsec fixture. Every issue is planted. DO NOT fix.
import hashlib
import os
import sqlite3
import subprocess


def sql_injection(username):
    conn = sqlite3.connect("app.db")
    cur = conn.cursor()
    # PLANT(py-sqli, min-profile=max, CWE-89): SQL injection via string formatting
    cur.execute("SELECT * FROM users WHERE name = '%s'" % username)
    return cur.fetchall()


def command_injection(host):
    # PLANT(py-cmdi, min-profile=standard, CWE-78): OS command injection via shell=True with concatenation
    return subprocess.check_output("ping -c 1 " + host, shell=True)


def weak_hash(password):
    # PLANT(py-weak-hash, min-profile=standard, CWE-327): weak hash (MD5) over a password (textbook CWE-328; semgrep emits CWE-327)
    return hashlib.md5(password.encode()).hexdigest()


def path_traversal(filename):
    # PLANT-GAP: path traversal via unsanitized user filename (CWE-22) — caught by no profile; tracked in docs/coverage.md
    with open(os.path.join("/var/data", filename)) as fh:
        return fh.read()
